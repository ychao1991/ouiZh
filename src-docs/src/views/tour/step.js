/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useState } from 'react';

import {
  OuiLink,
  OuiText,
  OuiSpacer,
  OuiTourStep,
} from '../../../../src/components';

export default () => {
  const [isOpen, setIsOpen] = useState(true);
  return (
    <div>
      <OuiTourStep
        content={
          <OuiText>
            <p>The tour step content.</p>
          </OuiText>
        }
        isStepOpen={isOpen}
        minWidth={300}
        onFinish={() => setIsOpen(false)}
        step={1}
        stepsTotal={1}
        title="Title of the current step"
        subtitle="Title of the full tour"
        anchorPosition="rightUp">
        <OuiText>
          The tour step{' '}
          <OuiLink onClick={() => setIsOpen(!isOpen)}>anchor point</OuiLink>.
        </OuiText>
      </OuiTourStep>
      <OuiSpacer size="xxl" />
      <OuiSpacer size="xxl" />
    </div>
  );
};
