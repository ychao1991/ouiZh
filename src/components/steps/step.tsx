/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import React, {
  createElement,
  FunctionComponent,
  HTMLAttributes,
  ReactNode,
} from 'react';
import { CommonProps } from '../common';
import { OuiTitle, OuiTitleProps, OuiTitleSize } from '../title';
import { OuiStepNumber, OuiStepStatus } from './step_number';

export interface OuiStepInterface {
  /**
   * ReactNode to render as this component's content
   */
  children: ReactNode;
  /**
   * The HTML tag used for the title
   */
  headingElement?: string;
  /**
   * The number of the step in the list of steps
   */
  step?: number;
  title: string;
  /**
   * May replace the number provided in props.step with alternate styling.
   */
  status?: OuiStepStatus;
  /**
   * Title sizing equivalent to OuiTitle, but only `m`, `s` and `xs`. Defaults to `s`
   */
  titleSize?: Exclude<OuiTitleProps['size'], 'xxxs' | 'xxs' | 'l'>;
}

export type OuiStepProps = CommonProps &
  Omit<HTMLAttributes<HTMLDivElement>, 'title'> &
  OuiStepInterface;

export const OuiStep: FunctionComponent<OuiStepProps> = ({
  className,
  children,
  headingElement = 'p',
  step = 1,
  title,
  titleSize = 's',
  status,
  ...rest
}) => {
  const classes = classNames(
    'ouiStep',
    {
      'ouiStep--small': titleSize === 'xs',
      'ouiStep-isDisabled': status === 'disabled',
    },
    className
  );
  const numberClasses = classNames('ouiStep__circle', {
    'ouiStepNumber--small': titleSize === 'xs',
  });

  return (
    <div className={classes} {...rest}>
      <div className="ouiStep__titleWrapper">
        <OuiStepNumber
          className={numberClasses}
          number={step}
          status={status}
          titleSize={titleSize}
          isHollow={status === 'incomplete'}
        />
        <OuiTitle size={titleSize as OuiTitleSize} className="ouiStep__title">
          {createElement(headingElement, null, title)}
        </OuiTitle>
      </div>

      <div className="ouiStep__content">{children}</div>
    </div>
  );
};
