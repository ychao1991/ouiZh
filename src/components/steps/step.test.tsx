/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { OuiStep } from './step';
import { STATUS } from './step_number';

describe('OuiStep', () => {
  test('is rendered', () => {
    const component = render(
      <OuiStep {...requiredProps} title={'First step'}>
        <p>Do this</p>
      </OuiStep>
    );

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    test('headingElement', () => {
      const component = render(
        <OuiStep headingElement={'h3'} title={'First step'}>
          <p>Do this</p>
        </OuiStep>
      );

      expect(component).toMatchSnapshot();
    });

    test('step', () => {
      const component = render(
        <OuiStep step={5} title={'First step'}>
          <p>Do this</p>
        </OuiStep>
      );

      expect(component).toMatchSnapshot();
    });

    test('titleSize', () => {
      const component = render(
        <OuiStep titleSize="xs" title={'First step'}>
          <p>Do this</p>
        </OuiStep>
      );

      expect(component).toMatchSnapshot();
    });

    describe('status', () => {
      STATUS.forEach((status) => {
        test(`${status} is rendered`, () => {
          const component = render(
            <OuiStep status={status} title={'First step'}>
              <p>Do this</p>
            </OuiStep>
          );

          expect(component).toMatchSnapshot();
        });
      });
    });
  });
});
