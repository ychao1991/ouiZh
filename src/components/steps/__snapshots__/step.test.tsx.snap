// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiStep is rendered 1`] = `
<div
  aria-label="aria-label"
  class="ouiStep testClass1 testClass2"
  data-test-subj="test subject string"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props headingElement 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <h3
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </h3>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status complete is rendered 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--complete ouiStep__circle"
    >
      <span
        aria-label="Step 1 is complete"
        class="ouiStepNumber__icon"
        data-ouiicon-type="check"
      />
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status danger is rendered 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--danger ouiStep__circle"
    >
      <span
        aria-label="Step 1 has errors"
        class="ouiStepNumber__icon"
        data-ouiicon-type="cross"
      />
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status disabled is rendered 1`] = `
<div
  class="ouiStep ouiStep-isDisabled"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--disabled ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is disabled
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status incomplete is rendered 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--incomplete ouiStepNumber-isHollow ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is incomplete
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status loading is rendered 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--loading ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1 is loading
      </span>
      <span
        class="ouiLoadingSpinner ouiLoadingSpinner--xLarge ouiStepNumber__loader"
      />
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props status warning is rendered 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStepNumber--warning ouiStep__circle"
    >
      <span
        aria-label="Step 1 has warnings"
        class="ouiStepNumber__icon"
        data-ouiicon-type="alert"
      />
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props step 1`] = `
<div
  class="ouiStep"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStep__circle"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 5
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        5
      </span>
    </span>
    <p
      class="ouiTitle ouiTitle--small ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;

exports[`OuiStep props titleSize 1`] = `
<div
  class="ouiStep ouiStep--small"
>
  <div
    class="ouiStep__titleWrapper"
  >
    <span
      class="ouiStepNumber ouiStep__circle ouiStepNumber--small"
    >
      <span
        class="ouiScreenReaderOnly"
      >
        Step 1
      </span>
      <span
        aria-hidden="true"
        class="ouiStepNumber__number"
      >
        1
      </span>
    </span>
    <p
      class="ouiTitle ouiTitle--xsmall ouiStep__title"
    >
      First step
    </p>
  </div>
  <div
    class="ouiStep__content"
  >
    <p>
      Do this
    </p>
  </div>
</div>
`;
